server/src/test/java/org/elasticsearch/index/snapshots/blobstore/BlobStoreIndexShardSnapshotsTest.java

```diff
@@ -0,0 +1,175 @@
+        // Method getThreadAllocatedBytes is a rough estimation and returned value fluctuates a little bit.
+        // Generate a bigger list so that difference with/without the fix is visible (like 3x) and can't be treated as a noise.
+        List<FileInfo> fileInfos = new ArrayList<>();
+        for (int i = 1; i <= 10_000; i++) {
+            String name = "name" + i;
+            fileInfos.add(
+                new FileInfo(
+                    name,
+                    new StoreFileMetadata(name, 1, "dummy_checksum", org.apache.lucene.util.Version.LUCENE_9_12_0),
+                    new ByteSizeValue(i)
+                )
+            );
+        }
+
+        // Different SnapshotFiles can refer to the same FileInfo-s in the SnapshotFiles.indexFiles field.
+        // Making 2 lists almost completely overlapping to make difference visible and eliminate getThreadAllocatedBytes fluctuation noise.
+        SnapshotFiles snapshot1 = new SnapshotFiles("1", fileInfos.subList(10, fileInfos.size()), null);
+        SnapshotFiles snapshot2 = new SnapshotFiles("1", fileInfos.subList(0, fileInfos.size() - 10), null);
+        return new BlobStoreIndexShardSnapshots(List.of(snapshot1, snapshot2));
+    }
+
+    private static boolean isSame(BlobStoreIndexShardSnapshots shardSnapshot1, BlobStoreIndexShardSnapshots shardSnapshot2) {
+        Map<String, FileInfo> files1 = shardSnapshot1.files();
+        Map<String, FileInfo> files2 = shardSnapshot1.files();
+        if (files1.size() != files2.size()) {
+            return false;
+        }
+        for (Map.Entry<String, FileInfo> entry : files1.entrySet()) {
+            if (entry.getValue().isSame(files2.get(entry.getKey())) == false) {
+                return false;
+            }
+        }
+
+        Map<String, List<FileInfo>> physicalFiles1 = shardSnapshot1.physicalFiles();
+        Map<String, List<FileInfo>> physicalFiles2 = shardSnapshot1.physicalFiles();
+        if (physicalFiles1.size() != physicalFiles2.size()) {
+            return false;
+        }
+        for (Map.Entry<String, List<FileInfo>> entry : physicalFiles1.entrySet()) {
+            List<FileInfo> list1 = entry.getValue();
+            List<FileInfo> list2 = physicalFiles2.get(entry.getKey());
+            if (list1.size() != list2.size()) {
+                return false;
+            }
+            for (int i = 0; i < list1.size(); i++) {
+                if (list1.get(i).isSame(list2.get(i)) == false) {
+                    return false;
+                }
+            }
+        }
+
+        List<SnapshotFiles> shapshotFiles1 = shardSnapshot1.snapshots();
+        List<SnapshotFiles> shapshotFiles2 = shardSnapshot1.snapshots();
+        if (shapshotFiles1.size() != shapshotFiles2.size()) {
+            return false;
+        }
+        for (int i = 0; i < shapshotFiles1.size(); i++) {
+            SnapshotFiles sf1 = shapshotFiles1.get(i);
+            SnapshotFiles sf2 = shapshotFiles2.get(i);
```

Comment:
`FileInfo` already has `isSame` method but otherwise had to go over all structures to check equality.

Didn't introduce `BlobStoreIndexShardSnapshots.equals` because we don't need in in production code and this is not something short to just add it there.